feature: choose custom install location

AlvinSchiller opened the pull request:
Added the possibility for the user to select a custom location for the game installation instead of only the drive root.

Considering the comment in https://github.com/MarcellVokk/aio-launcher/issues/3 this is available via a "advenced settings" toggleswitch in the installation popup. The original selection is still default for the straight forward case.

Update:
- added toggle switch to change to folder selection settings
- set "program files (x86)" as defaultpath for folder selection
- folders containing ":\Windows" or "\OneDrive" are excluded
- valid drives have to be of type "fixed" (no removable or network drives) -> also effects the original selection
- if no drive is available an error message is shown


<img width="673" height="394" alt="1_Original_selection" src="https://github.com/user-attachments/assets/5da98ca9-b9b7-457d-a5b9-6331c692504c" />
<img width="671" height="395" alt="2_Folder_selection_with_defaultpath" src="https://github.com/user-attachments/assets/ea831d7d-8f55-4b23-bd6a-4bc938c0e5d4" />
<img width="670" height="393" alt="3_Folder_selection_with_custompath" src="https://github.com/user-attachments/assets/8b1b5f1a-b49f-4be6-934d-5080fb169dd9" />
<img width="677" height="482" alt="4_Folder_selection_dialog" src="https://github.com/user-attachments/assets/d93fc984-2c63-4def-a186-9f4f5c5918ea" />
<img width="682" height="402" alt="5_with_error" src="https://github.com/user-attachments/assets/ef3fa4f4-f748-4839-8175-dd49781ab440" />

AlvinSchiller committed:
feat: added FolderPicker for installation popup to choose installation location

AlvinSchiller committed:
fix: Move FolderPicker

AlvinSchiller committed:
fix: refactored code to DriveUtils class

AlvinSchiller committed:
fix: update folderpicker

AlvinSchiller committed:
feat: Location selektion updates.
Use old behavior as default, but make advanced selektion possible

AlvinSchiller committed:
feat: update layout and simplify error handling

AlvinSchiller committed:
fix: cleanup

AlvinSchiller committed:
fix: cleanup

AlvinSchiller committed:
fix: Revert: introduce in another pr

AlvinSchiller committed:
fix: use existing toggleable element. align style to system settings toggleable

MarcellVokk commented on src/BfmeFoundationProject_AllInOneLauncher/Elements/Native/FolderPicker.cs:
Is this necesary? Use Microsoft.Win32.OpenFolderDialog instead, it's built in and as a bonus it's the new style folder browser, not the old style crappy one.

MarcellVokk reviewed (changes requested):
Hi! Made a few comments on the code that need to be worked on before I add this to the main branch.

Also, the UI doesn't look quite right to me, that toggle is placed a bit akwardly, maybe put it inline with the "Location" title, or keep it where it is but put a background on it, or have the toggle text start where the Location title starts horizontaly.
And when "advanced mode" is turned on, there is nothing to indicate that you are supposed to click on the box below to select a folder, maybe put a little folder icon on the right side of the box, and also reduce the vertical padding on that box because with 2 lines it looks a bit akward with that much padding.

AlvinSchiller committed:
fix: use Win32 FolderDialog

AlvinSchiller committed:
fix: update variable names

AlvinSchiller committed:
fix: update subelement style

AlvinSchiller committed:
fix: set HiglightButton style for folderselection button

AlvinSchiller committed:
fix: Update advanced folder selection ui

use visiblity hidden instead of collapse to avoid layout shift
reduce layout complexity by removing obsolete elements

MarcellVokk commented:
Are you kidding me? That UI looks like shit...

AlvinSchiller committed:
fix: Update advanced folder selection ui. refactor to single button.

AlvinSchiller commented:
@MarcellVokk 
> Are you kidding me? That UI looks like shit...

Do you mean the folder selection or also the change of the indentation?

Regarding the folder selection i reverted this to the single button design and added the folder icon.

<img width="669" height="390" alt="Screenshot 2025-12-28 230245" src="https://github.com/user-attachments/assets/84722d01-d052-493d-b389-43e543418c35" />
<img width="671" height="391" alt="Screenshot 2025-12-28 230401" src="https://github.com/user-attachments/assets/772f4f2f-82fd-4e1f-b54a-c9ae0107068f" />

AlvinSchiller commented:
Regarding your earlier statement 

>  ... and also reduce the vertical padding on that box because with 2 lines it looks a bit akward with that much padding.

For now i used the default padding of the "HiglightButtonStyle" to keep a consistent look.
Also, in my eyes, it looks odd if the vertical padding is reduced. But your call which way to go.
Here an example if the vertical padding is halfed (from 10 to 5)

<img width="671" height="393" alt="Screenshot 2025-12-28 230641" src="https://github.com/user-attachments/assets/6076f6db-ad62-4f0a-8f17-10b7364b3d65" />
<img width="674" height="393" alt="Screenshot 2025-12-28 230704" src="https://github.com/user-attachments/assets/4a9463d6-b4e8-480d-856d-93ab4b176fd2" />

inmate37 commented:
@AlvinSchiller @MarcellVokk hey guys, can this one get a little attention ?

lfechne commented:
That UI looks okay for a home made app. If UI is your only concern and not security or unexpected behavior of the PR code, then just accept the PR.

> Are you kidding me? That UI looks like shit...

Are you kidding us? That is not respectable language towards a fellow developer! You should be ashamed.
If you want to bring together a community here, then be a respectable and grateful person for any helpers.

inmate37 commented:
@MarcellVokk The UI can be changed later if necessary, as for the code it is okay, lets merge this asap, required feature

inmate37 commented:
@MarcellVokk If you are busy and cannot review this please consider adding new reviewers for this one

AlvinSchiller commented:
As a workaround to change an already existing installation to a custom path  (e.g. "D:\BFME2" -> "D:\Games\BFME\BFME2"):
(I did this myself before i created this pr. This is nothing official and at own risk)

- close the launcher
- open "regedit" as an administrator
- search for the current installation path of the game
- change all entries of the game to the new custom path
- move the game folder from the old to the new path
 - open the launcher
 - validate that the games is still recognized as installed (if not check the paths again)

inmate37 commented:
Since @MarcellVokk doesn't give 2 F's about our concern, use the following guide to make requested feature possible:

**1.** 
Create the folder `Games` in drive `D:`

**PS:** _You can set whatever name, and install the game in whatever drive, I will use drive `D:` and folder name `Games` as an example_

**2.** 
Move the folders from:
```
D:\BFME2
D:\BFME Workshop
D:\RotWK
```
To:
```
D:\Games\BFME2
D:\Games\BFME Workshop
D:\Games\RotWK
```

**3.** 
Open: `Win+R`, then type: `regedit`

Go to:
`HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Electronic Arts\Electronic Arts\The Battle for Middle-earth II`

`HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Electronic Arts\Electronic Arts\The Lord of the Rings, The Rise of the Witch-king`

Change the value in `InstallPath` to: `D:\Games\...`

**4.** 
Go to:
`HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths\lotrbfme2.exe`
`HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths\lotrbfme2ep1.exe`

Change the value in `Default` and `Path` to: `D:\Games\...`

**5.** 
Run `AllInOneLauncher.exe` and check that the game is set to: `Play` not `Install`
